### PR TITLE
Omitted the walks in visit_expr() and visit_stmt()

### PR DESCRIPTION
**Pull request by TheDarkula**
@oli-obk

### CODE REVIEW

**Comment by oli-obk** [comment]
Hmm... curious, it fails on paths. Try adding `visitor.visit_qpath(qpath, expression.id, expression.span);` to the `ExprPath` arm.

File: src/librustc_passes/rvalue_promotion.rs [hunk 1]
```diff
@@ -445,40 +485,141 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
                 }
             }
         }
 
-        hir::ExprBlock(..) |
-        hir::ExprIndex(..) |
-        hir::ExprArray(_) |
-        hir::ExprType(..) |
-        hir::ExprTup(..) => {}
+        hir::ExprBlock(ref box_block, ref _option_label) => {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: Aha! You also need to visit all statements inside the block, not just the `expr`

calling `visit_stmt` on them should suffice.

File: src/librustc_passes/rvalue_promotion.rs [hunk 2]
```diff
@@ -445,40 +485,141 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+        }
+
+        hir::ExprIndex(ref lhs, ref rhs) => {
+            v.visit_expr(lhs);
+            v.visit_expr(rhs);
+        }
+
+        hir::ExprArray(ref hirvec) => {
+            for index in hirvec.iter() {
+                match *index {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: you don't need this `match`, it should be fine to call `v.visit_expr(index)` directly, if that fails, use `v.visit_expr(&**index)` to force a deref.

That transformation can be applied to all the `match` statements you used for reference conversions

File: src/librustc_passes/rvalue_promotion.rs [hunk 3]
```diff
@@ -293,27 +293,38 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
         hir::ExprIndex(..) if v.tables.is_method_call(e) => {
             v.promotable = false;
         }
-        hir::ExprBox(_) => {
+        hir::ExprBox(ref expr) => {
+            match *expr {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: this match is also unnecessary

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc_passes/rvalue_promotion.rs [hunk 4]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+            v.visit_expr(&expr)
+        }
 
         // Expressions with side-effects.
-        hir::ExprAssign(..) |
-        hir::ExprAssignOp(..) |
-        hir::ExprInlineAsm(..) => {
+        hir::ExprAssignOp(_, ref lhs, ref rhs) | hir::ExprAssign(ref lhs, ref rhs) => {
+            v.visit_expr(lhs);
+            v.visit_expr(rhs);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: you lost the `v.promotable = false;` here

File: src/librustc_passes/rvalue_promotion.rs [hunk 5]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+                None => {},
+            }
+        }
+
+        hir::ExprContinue(_) => {}
 
         // Generator expressions
-        hir::ExprYield(_) |
+        hir::ExprYield(ref expr) => {
+            v.visit_expr(&expr)
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: `v.promotable = false;` missing here

File: src/librustc_passes/rvalue_promotion.rs [hunk 6]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
-        hir::ExprContinue(_) |
-        hir::ExprRet(_) |
+        hir::ExprBreak(_, ref option_expr) | hir::ExprRet(ref option_expr) => {
+            match *option_expr {
+                Some(ref expr) => { v.visit_expr(&expr) },
+                None => {},
+            }
+        }
+
+        hir::ExprContinue(_) => {}
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: you lost the `v.promotable = false;` here

File: src/librustc_passes/rvalue_promotion.rs [hunk 7]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+
+        hir::ExprLoop(ref box_block, ref _option_label, ref _loop_source) => {
+            v.visit_block(box_block);
+        }
 
         // More control flow (also not very meaningful).
-        hir::ExprBreak(..) |
-        hir::ExprContinue(_) |
-        hir::ExprRet(_) |
+        hir::ExprBreak(_, ref option_expr) | hir::ExprRet(ref option_expr) => {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: `v.promotable = false;` missing here

File: src/librustc_passes/rvalue_promotion.rs [hunk 8]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
         // Loops (not very meaningful in constants).
-        hir::ExprWhile(..) |
-        hir::ExprLoop(..) |
+        hir::ExprWhile(ref expr, ref box_block, ref _option_label) => {
+            v.visit_expr(expr);
+            v.visit_block(box_block);
+        }
+
+        hir::ExprLoop(ref box_block, ref _option_label, ref _loop_source) => {
+            v.visit_block(box_block);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: `v.promotable = false;` missing here

File: src/librustc_passes/rvalue_promotion.rs [hunk 9]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+                Some(ref expr) => v.visit_expr(&expr),
+                None => {},
+            }
+        }
 
         // Loops (not very meaningful in constants).
-        hir::ExprWhile(..) |
-        hir::ExprLoop(..) |
+        hir::ExprWhile(ref expr, ref box_block, ref _option_label) => {
+            v.visit_expr(expr);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: `v.promotable = false;` missing here

File: src/librustc_passes/rvalue_promotion.rs [hunk 10]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+            }
+        }
+
+        hir::ExprIf(ref lhs, ref rhs, ref option_expr) => {
+            v.visit_expr(lhs);
+            v.visit_expr(rhs);
+            match option_expr {
+                Some(ref expr) => v.visit_expr(&expr),
+                None => {},
+            }
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: `v.promotable = false;` missing here

File: src/librustc_passes/rvalue_promotion.rs [hunk 11]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+                    ref arm => {
+                        match arm.body {
+                            ref expr => v.visit_expr(&*expr),
+                        }
+                        match arm.guard {
+                        Some(ref expr) => v.visit_expr(&expr),
+                        None => {},
+                    } }
+                }
+            }
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: `v.promotable = false;` missing here

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc_passes/rvalue_promotion.rs [hunk 12]
```diff
@@ -271,6 +269,19 @@ impl<'a, 'tcx> Visitor<'tcx> for CheckCrateVisitor<'a, 'tcx> {
...
+    fn visit_block(&mut self, box_block: &'tcx hir::Block) {
+        match *box_block {
+            ref block => {
+                for index in block.stmts.iter() {
+                    self.visit_stmt(index)
+                }
+                match block.expr {
+                    Some(ref box_expr) => { self.visit_expr(&*box_expr) },
+                    None => {},
+                } }
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: can you add a line-break between those two closing bracket?

File: src/librustc_passes/rvalue_promotion.rs [hunk 13]
```diff
@@ -271,6 +269,19 @@ impl<'a, 'tcx> Visitor<'tcx> for CheckCrateVisitor<'a, 'tcx> {
         }
         self.promotable &= outer;
     }
+
+    fn visit_block(&mut self, box_block: &'tcx hir::Block) {
+        match *box_block {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: rename argument to `block` and remove the match

File: src/librustc_passes/rvalue_promotion.rs [hunk 14]
```diff
@@ -288,32 +301,36 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
-            v.promotable = false;
-        }
-        hir::ExprBox(_) => {
+        hir::ExprBox(ref expr) => {
+            v.visit_expr(&expr);
             v.promotable = false;
         }
-        hir::ExprUnary(op, _) => {
+        hir::ExprUnary(op, ref expr) => {
+            match *expr {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: still a nop-match left

File: src/librustc_passes/rvalue_promotion.rs [hunk 15]
```diff
@@ -288,32 +301,36 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+                        op.node == hir::BiLe || op.node == hir::BiLt ||
+                        op.node == hir::BiGe || op.node == hir::BiGt);
 
                     v.promotable = false;
                 }
                 _ => {}
             }
         }
         hir::ExprCast(ref from, _) => {
+            match *from {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: another nop-match

File: src/librustc_passes/rvalue_promotion.rs [hunk 16]
```diff
@@ -379,7 +397,13 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
                 }
             }
         }
-        hir::ExprCall(ref callee, _) => {
+        hir::ExprCall(ref callee, ref hirvec) => {
+            match **callee {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: and another

File: src/librustc_passes/rvalue_promotion.rs [hunk 17]
```diff
@@ -424,7 +451,16 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
                 v.tcx.sess.delay_span_bug(e.span, "no type-dependent def for method call");
             }
         }
-        hir::ExprStruct(..) => {
+        hir::ExprStruct(ref _qpath, ref hirvec, ref option_expr) => {
+            for index in hirvec.iter() {
+                match *index {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: also here

File: src/librustc_passes/rvalue_promotion.rs [hunk 18]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
             // as there are none, and thus closures can't have upvars there.
             if v.tcx.with_freevars(e.id, |fv| !fv.is_empty()) {
                 v.promotable = false;
             }
         }
 
-        hir::ExprField(ref expr, _) => {
+        hir::ExprField(ref expr, _ident) => {
+            match *expr {
+                ref expr => { v.visit_expr(&expr) },
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: and here

File: src/librustc_passes/rvalue_promotion.rs [hunk 19]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+        }
+
 
         // Conditional control flow (possible to implement).
-        hir::ExprMatch(..) |
-        hir::ExprIf(..) |
+        hir::ExprMatch(ref expr, ref hirvec_arm, ref _match_source) => {
+            v.visit_expr(expr);
+            for index in hirvec_arm.iter() {
+                match *index {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: remove nop match

File: src/librustc_passes/rvalue_promotion.rs [hunk 20]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+                    ref arm => {
+                        match arm.body {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: remove this nop-match, too. just use `v.visit_expr(&arm.body)`

File: src/librustc_passes/rvalue_promotion.rs [hunk 21]
```diff
@@ -433,52 +469,130 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+                            ref expr => v.visit_expr(&*expr),
+                        }
+                        match arm.guard {
+                        Some(ref expr) => v.visit_expr(&expr),
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: fix indentation and newlines here

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: src/librustc_passes/rvalue_promotion.rs [hunk 22]
```diff
@@ -288,32 +298,36 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
-                            op.node == hir::BiLe || op.node == hir::BiLt ||
-                            op.node == hir::BiGe || op.node == hir::BiGt);
+                        op.node == hir::BiLe || op.node == hir::BiLt ||
+                        op.node == hir::BiGe || op.node == hir::BiGt);
 
                     v.promotable = false;
                 }
                 _ => {}
             }
+            v.promotable = false;
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: uhm... what is this line doing here?

File: src/librustc_passes/rvalue_promotion.rs [hunk 23]
```diff
@@ -288,32 +298,36 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+            v.visit_expr(&expr);
             v.promotable = false;
         }
-        hir::ExprUnary(op, _) => {
+        hir::ExprUnary(op, ref expr) => {
+            v.visit_expr(expr);
             if op == hir::UnDeref {
                 v.promotable = false;
             }
+            v.promotable = false;
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: and this one?

File: src/librustc_passes/rvalue_promotion.rs [hunk 24]
```diff
@@ -288,32 +286,36 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+                        op.node == hir::BiGe || op.node == hir::BiGt);
 
                     v.promotable = false;
                 }
                 _ => {}
             }
+            v.promotable = false;
         }
         hir::ExprCast(ref from, _) => {
+            match *from {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: still a nop-match left

File: src/librustc_passes/rvalue_promotion.rs [hunk 25]
```diff
@@ -433,52 +450,148 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+            v.visit_block(box_block);
+        }
+
+        hir::ExprIndex(ref lhs, ref rhs) => {
+            if v.tables.is_method_call(e) {
+                v.promotable = false;
+            }
+            v.visit_expr(lhs);
+            v.visit_expr(rhs);
+            v.promotable = false;
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: This `v.promotable = false` is wrong

File: src/librustc_passes/rvalue_promotion.rs [hunk 26]
```diff
@@ -231,12 +233,12 @@ impl<'a, 'tcx> Visitor<'tcx> for CheckCrateVisitor<'a, 'tcx> {
                     hir::DeclItem(_) => {}
                 }
             }
-            hir::StmtExpr(..) |
-            hir::StmtSemi(..) => {
-                self.promotable = false;
+            hir::StmtExpr(ref box_expr, node_id) |
+            hir::StmtSemi(ref box_expr, node_id) => {
+                self.visit_expr(box_expr);
+                self.visit_id(node_id);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: You lost the `self.promotable = false`

Comment (oli-obk) [reply src/librustc_passes/rvalue_promotion.rs]: Alsono need to visit the id

File: src/librustc_passes/rvalue_promotion.rs [hunk 27]
```diff
@@ -216,7 +216,9 @@ impl<'a, 'tcx> Visitor<'tcx> for CheckCrateVisitor<'a, 'tcx> {
 
     fn visit_stmt(&mut self, stmt: &'tcx hir::Stmt) {
         match stmt.node {
-            hir::StmtDecl(ref decl, _) => {
+            hir::StmtDecl(ref decl, node_id) => {
+                self.visit_id(node_id);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: No need to visit the ID, and probably not the decl either

File: src/librustc_passes/rvalue_promotion.rs [hunk 28]
```diff
@@ -288,32 +288,32 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+        hir::ExprBox(ref expr) => {
+            v.visit_expr(&expr);
             v.promotable = false;
         }
-        hir::ExprBox(_) => {
-            v.promotable = false;
-        }
-        hir::ExprUnary(op, _) => {
+        hir::ExprUnary(op, ref expr) => {
+            v.visit_expr(expr);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: Still missing the `if v.tables.is_method_call(e) { v.promotable = false; }`

File: src/librustc_passes/rvalue_promotion.rs [hunk 29]
```diff
@@ -288,32 +288,32 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
             if op == hir::UnDeref {
                 v.promotable = false;
             }
         }
-        hir::ExprBinary(op, ref lhs, _) => {
+        hir::ExprBinary(op, ref lhs, ref rhs) => {
+            v.visit_expr(lhs);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: Same here wrt method call

File: src/librustc_passes/rvalue_promotion.rs [hunk 30]
```diff
@@ -324,6 +324,7 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
             }
         }
         hir::ExprPath(ref qpath) => {
+            v.visit_qpath(qpath, e.id, e.span);
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: Not necessary

File: src/librustc_passes/rvalue_promotion.rs [hunk 31]
```diff
@@ -433,52 +448,146 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
-        hir::ExprLit(_) |
-        hir::ExprAddrOf(..) |
-        hir::ExprRepeat(..) => {}
+        hir::ExprLit(_) => {}
+
+        hir::ExprAddrOf(ref _mutability, ref expr) => {
+            v.visit_expr(expr);
+        }
+
+        hir::ExprRepeat(ref expr, ref _anon_cast) => {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: Merge with `ExprAddrOf`

File: src/librustc_passes/rvalue_promotion.rs [hunk 32]
```diff
@@ -433,52 +448,146 @@ fn check_expr<'a, 'tcx>(v: &mut CheckCrateVisitor<'a, 'tcx>, e: &hir::Expr, node
...
+            for pat in hirvec_arm.iter().flat_map(|arm| &arm.pats) {
+                mut_borrow = v.remove_mut_rvalue_borrow(pat);
+            }
+            if mut_borrow {
+                v.mut_rvalue_borrows.insert(expr.id);
+            }
+
+            v.visit_expr(expr);
+            for index in hirvec_arm.iter() {
+                match *index {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: Nop match

File: src/librustc_passes/rvalue_promotion.rs [hunk 33]
```diff
@@ -169,12 +168,7 @@ impl<'a, 'gcx> CheckCrateVisitor<'a, 'gcx> {
     }
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for CheckCrateVisitor<'a, 'tcx> {
-    fn nested_visit_map<'this>(&'this mut self) -> NestedVisitorMap<'this, 'tcx> {
-        // note that we *do* visit nested bodies, because we override `visit_nested_body` below
-        NestedVisitorMap::None
-    }
-
+impl<'a, 'tcx>CheckCrateVisitor<'a, 'tcx> {
```
Comment (oli-obk) [inline src/librustc_passes/rvalue_promotion.rs]: Style nit: add a space before the `Check`

**Review by oli-obk — APPROVED** [review]

**Comment by RalfJung** [comment]
Looks like a spurious network failure.

File: src/librustc_passes/rvalue_promotion.rs [hunk 34]
```diff
@@ -169,12 +168,7 @@ impl<'a, 'gcx> CheckCrateVisitor<'a, 'gcx> {
     }
 }
 
-impl<'a, 'tcx> Visitor<'tcx> for CheckCrateVisitor<'a, 'tcx> {
```
Comment (eddyb) [inline src/librustc_passes/rvalue_promotion.rs]: I should note that this part of the change *makes no sense* without #52318.
Please motivate such changes better in the future, using the PR description.

**Pull request merged**